Framework/HistogramData/src/HistogramItem.cpp

```diff
@@ -0,0 +1,106 @@
+#include "MantidHistogramData/HistogramItem.h"
+#include "MantidHistogramData/Histogram.h"
+#include <algorithm>
+#include <utility>
+
+using Mantid::HistogramData::HistogramItem;
+using Mantid::HistogramData::Histogram;
+using Mantid::HistogramData::BinEdges;
+using Mantid::HistogramData::Points;
+
+HistogramItem::HistogramItem(const Histogram &histogram, const size_t index)
+    : m_histogram(histogram), m_index(index) {}
+
+double HistogramItem::counts() const {
+  const auto &yMode = histogramRef().yMode();
+  if (yMode == Histogram::YMode::Counts) {
+    return histogramRef().counts()[m_index];
+  } else {
+    const Frequencies frequency{histogramRef().frequencies()[m_index]};
+    return Counts(frequency, binEdges())[0];
+  }
+}
+
+double HistogramItem::countVariance() const {
+  return histogramRef().countVariances()[m_index];
+}
+
+double HistogramItem::countStandardDeviation() const {
+  return histogramRef().countStandardDeviations()[m_index];
+}
+
+double HistogramItem::frequency() const {
+  const auto &yMode = histogramRef().yMode();
+  if (yMode == Histogram::YMode::Frequencies) {
+    return histogramRef().frequencies()[m_index];
+  } else {
+    const Counts counts{histogramRef().counts()[m_index]};
+    return Frequencies(counts, binEdges())[0];
+  }
+}
+
+double HistogramItem::frequencyVariance() const {
+  return histogramRef().frequencyVariances()[m_index];
+}
+
+double HistogramItem::frequencyStandardDeviation() const {
+  return histogramRef().frequencyStandardDeviations()[m_index];
+}
+
+double HistogramItem::center() const { return point()[0]; }
+
+double HistogramItem::width() const {
+  const auto &edges = binEdges();
+  const auto &lower = edges[0];
+  const auto &upper = edges[1];
+  return upper - lower;
+}
+
+const BinEdges HistogramItem::binEdges() const {
+  const auto &xMode = histogramRef().xMode();
+
+  if (xMode == Histogram::XMode::BinEdges) {
+    const auto &edges = histogramRef().binEdges();
+    const auto &lower = edges[m_index];
+    const auto &upper = edges[m_index + 1];
+    return BinEdges{lower, upper};
+  } else {
+    const auto &points = histogramRef().points();
+    // Guess the bin edges from the points
+    if (points.size() == 0) {
```

Comment:
In that case `histogram.begin() == histogram.end()`, i.e., this should never happen.

Reply:
Writing a test for the case that `histogram.begin() == histogram.end()` it seems that `histogram.size()` does not correctly report `0`. For example, the following fails:
```c++
    Histogram hist(Histogram::XMode::BinEdges, Histogram::YMode::Counts);
    assert(hist.size() == 0);
```

Reply:
Actually I fixed that yesterday in https://github.com/mantidproject/mantid/tree/21128_ExtractSpectra_refactor since it also causes trouble there. In general this cannot (could not) happen in Mantid because `MatrixWorkspace` does not allow creation with X or Y length 0, so I had not paid much attention to that case. However, with support for non-constant histogram size added recently this becomes more relevant.

I will extract the relevant changes from that branch and make it a separate PR (since the bigger changes in the aforementioned branch will take slightly longer to get into `master`).

Reply:
@samueljackson92 I have assigned https://github.com/mantidproject/mantid/pull/21153 to you.

Reply:
Awesome. Thanks.